### PR TITLE
feature(zjow): add qgpo policy for new DI-engine pipeline

### PR DESCRIPTION
**Pull request by zjowowen**
Add qgpo policy for new DI-engine pipeline based on the implementation of repo [https://github.com/ChenDRAG/CEP-energy-guided-diffusion](url).

### CODE REVIEW

File: ding/model/common/encoder.py [hunk 1]
```diff
@@ -470,3 +471,17 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
         if self.final_relu:
             x = torch.relu(x)
         return x
+
+
+class GaussianFourierProjectionTimeEncoder(nn.Module):
```
Comment (PaParaZz1) [inline ding/model/common/encoder.py:476]: polish comments add add unittest

File: ding/policy/qgpo.py [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+
+
+@POLICY_REGISTRY.register('qgpo')
+class QGPOPolicy(Policy):
+    '''
+       Overview:
+            Policy class of QGPO algorithm
+            Contrastive Energy Prediction for Exact Energy-Guided Diffusion Sampling in Offline Reinforcement Learning
+            https://arxiv.org/abs/2304.12824
+    '''
```
Comment (PaParaZz1) [inline ding/policy/qgpo.py:24]: """

File: ding/policy/qgpo.py [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+        # (bool) Whether to use cuda for network.
+        cuda=False,
+        # (bool type) on_policy: Determine whether on-policy or off-policy.
+        # on-policy setting influences the behaviour of buffer.
+        # Default False in QGPO.
+        on_policy=False,
+        multi_agent=False,
+        model=dict(
+            score_net=dict(
+                score_base=dict(
```
Comment (PaParaZz1) [inline ding/policy/qgpo.py:38]: just use one-layer dict here

File: ding/policy/qgpo.py [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+        return {i: {"action": d} for i, d in zip(data_id, output)}
+
+    def _get_train_sample(self) -> None:
+        pass
+
+    def _process_transition(self) -> None:
+        pass
+
+    def _state_dict_learn(self) -> Dict[str, Any]:
+        ret = {
```
Comment (PaParaZz1) [inline ding/policy/qgpo.py:169]: directly return

File: ding/torch_utils/network/activation.py [hunk 5]
```diff
@@ -87,6 +87,15 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
         return 0.5 * x * (1.0 + torch.tanh(math.sqrt(2.0 / math.pi) * (x + 0.044715 * torch.pow(x, 3.0))))
 
 
+class SiLU(nn.Module):
```
Comment (PaParaZz1) [inline ding/torch_utils/network/activation.py:90]: why not use torch.SiLU

File: dizoo/d4rl/config/halfcheetah_medium_expert_qgpo_config.py [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+            learning_rate=1e-4,
+            batch_size=4096,
+            batch_size_q=256,
+            M=16,
+            diffusion_steps=15,
+            behavior_policy_stop_training_iter=600000,
+            energy_guided_policy_begin_training_iter=600000,
+            q_value_stop_training_iter=1100000,
+        ),
+        collect=dict(unroll_len=1, ),
```
Comment (PaParaZz1) [inline dizoo/d4rl/config/halfcheetah_medium_expert_qgpo_config.py:41]: remove this field

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: ding/example/qgpo.py [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+from ding.framework.context import OfflineRLContext
+from ding.framework.middleware import trainer, CkptSaver, offline_logger, wandb_offline_logger, termination_checker
+from ding.framework.middleware.functional.evaluator import qgpo_interaction_evaluator
+from ding.framework.middleware.functional.data_processor import qgpo_support_data_generator, qgpo_offline_data_fetcher
+from ding.utils import set_pkg_seed
+
+from dizoo.d4rl.config.halfcheetah_medium_expert_qgpo_config import main_config, create_config
+
+
+class QGPO_D4RLDataset(torch.utils.data.Dataset):
```
[the remark this answers is not in the document]
Comment (zjowowen) [reply ding/example/qgpo.py]: fixed.

File: ding/model/common/encoder.py (hunk 1 above)
Comment (zjowowen) [reply ding/model/common/encoder.py:476]: fixed.

File: ding/policy/qgpo.py (hunk 2 above)
Comment (zjowowen) [reply ding/policy/qgpo.py]: fixed.

File: ding/policy/qgpo.py (hunk 3 above)
Comment (zjowowen) [reply ding/policy/qgpo.py]: fixed.

File: ding/policy/qgpo.py (hunk 4 above)
Comment (zjowowen) [reply ding/policy/qgpo.py]: fixed.

File: ding/torch_utils/network/activation.py (hunk 5 above)
Comment (zjowowen) [reply ding/torch_utils/network/activation.py]: fixed.

File: dizoo/d4rl/config/halfcheetah_medium_expert_qgpo_config.py (hunk 6 above)
Comment (zjowowen) [reply dizoo/d4rl/config/halfcheetah_medium_expert_qgpo_config.py:41]: fixed.

File: ding/example/qgpo.py [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+            's_': self.next_states[index % self.len],
+            'd': self.is_finished[index % self.len],
+            'fake_a': self.fake_actions[index % self.len]
+            if hasattr(self, "fake_actions") else 0.0,  # self.fake_actions <D, 16, A>
+            'fake_a_': self.fake_next_actions[index % self.len]
+            if hasattr(self, "fake_next_actions") else 0.0,  # self.fake_next_actions <D, 16, A>
+        }
+        return data
+
+    def __add__(self, other):
```
Comment (PaParaZz1) [inline ding/example/qgpo.py:65]: why add this method

File: ding/policy/qgpo.py [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+
+    def _init_eval(self) -> None:
+        self.guidance_scale = self._cfg.eval.guidance_scale
+        self.diffusion_steps = self._cfg.eval.diffusion_steps
+
+    def _forward_eval(self, data: dict) -> dict:
+        data_id = list(data.keys())
+        data = default_collate(list(data.values()))
+        states = data
+        actions = self._model.score_model.select_actions(states, diffusion_steps=self.diffusion_steps)
```
Comment (PaParaZz1) [inline ding/policy/qgpo.py:150]: why not use `data` directly

File: dizoo/d4rl/config/halfcheetah_medium_expert_qgpo_config.py [hunk 10]
```diff
@@ -0,0 +1,56 @@
+from easydict import EasyDict
+
+main_config = dict(
+    exp_name='halfcheetah_medium_expert_v2_QGPO_seed0',
+    seed=0,
```
Comment (PaParaZz1) [inline dizoo/d4rl/config/halfcheetah_medium_expert_qgpo_config.py:5]: remove seed here

File: ding/torch_utils/network/res_block.py [hunk 11]
```diff
@@ -148,3 +148,21 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
         if self.dropout is not None:
             x = self.dropout(x)
         return x
+
+
+class TemporalSpatialResBlock(nn.Module):
```
Comment (PaParaZz1) [inline ding/torch_utils/network/res_block.py:153]: add usage comments for this class, add typing lint

File: ding/torch_utils/network/res_block.py [hunk 12]
```diff
@@ -148,3 +148,21 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
+
+    def __init__(self, input_dim, output_dim, t_dim=128, activation=torch.nn.SiLU):
```
Comment (PaParaZz1) [inline ding/torch_utils/network/res_block.py:155]: why not use `torch.nn.SiLU()`

File: ding/torch_utils/network/res_block.py [hunk 13]
```diff
@@ -148,3 +148,21 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
+        super().__init__()
+        self.time_mlp = nn.Sequential(
+            activation(),
```
Comment (PaParaZz1) [inline ding/torch_utils/network/res_block.py:158]: why first use activation here, add some comments

File: ding/policy/qgpo.py [hunk 14]
```diff
@@ -0,0 +1,172 @@
...
+    )
+
+    def _init_learn(self) -> None:
+        self.cuda = self._cfg.cuda
+        if self.cuda:
+            self.margin_prob_std_fn = functools.partial(marginal_prob_std, device=self._device)
+        self.behavior_model_optimizer = torch.optim.Adam(
+            self._model.score_model.parameters(), lr=self._cfg.learn.learning_rate
+        )
+        self.behavior_policy_stop_training_iter = self._cfg.learn.behavior_policy_stop_training_iter if hasattr(
```
Comment (PaParaZz1) [inline ding/policy/qgpo.py:84]: why do we use if expression here, we have already set this field in default config

File: ding/policy/qgpo.py [hunk 15]
```diff
@@ -0,0 +1,172 @@
...
+            hasattr(self._cfg.learn, 'energy_guided_policy_begin_training_iter') else 0
+        self.q_value_stop_training_iter = self._cfg.learn.q_value_stop_training_iter if hasattr(
+            self._cfg.learn, 'q_value_stop_training_iter'
+        ) and self._cfg.learn.q_value_stop_training_iter >= 0 else np.inf
+
+    def _forward_learn(self, data: dict) -> Dict[str, Any]:
+        if self.cuda:
+            data = {k: d.to(self._device) for k, d in data.items()}
+        else:
+            data = {k: d for k, d in data.items()}
```
Comment (PaParaZz1) [inline ding/policy/qgpo.py:97]: why do we need this code line

File: ding/policy/qgpo.py [hunk 16]
```diff
@@ -0,0 +1,172 @@
...
+        ) else np.inf
+        self.energy_guided_policy_begin_training_iter = self._cfg.learn.energy_guided_policy_begin_training_iter if \
+            hasattr(self._cfg.learn, 'energy_guided_policy_begin_training_iter') else 0
+        self.q_value_stop_training_iter = self._cfg.learn.q_value_stop_training_iter if hasattr(
+            self._cfg.learn, 'q_value_stop_training_iter'
+        ) and self._cfg.learn.q_value_stop_training_iter >= 0 else np.inf
+
+    def _forward_learn(self, data: dict) -> Dict[str, Any]:
+        if self.cuda:
+            data = {k: d.to(self._device) for k, d in data.items()}
```
Comment (PaParaZz1) [inline ding/policy/qgpo.py:95]: from ding.torch_utils import to_device

File: ding/model/common/encoder.py [hunk 17]
```diff
@@ -470,3 +471,46 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
+        Shapes:
+            - x (:obj:`torch.Tensor`): :math:`(B,)`, where B is batch size.
+            - output (:obj:`torch.Tensor`): :math:`(B, embed_dim)`, where B is batch size, embed_dim is the \
+                dimension of the output embedding vector.
+        Examples:
+            >>> encoder = GaussianFourierProjectionTimeEncoder(128)
+            >>> x = torch.randn(100)
+            >>> output = encoder(x)
+        """
+        x_proj = x[..., None] * self.W[None, :] * 2 * np.pi
```
Comment (PaParaZz1) [inline ding/model/common/encoder.py:515]: can we pre-compute `self.W[None, :] * 2 * np.pi`

File: ding/model/template/qgpo.py [hunk 18]
```diff
@@ -0,0 +1,365 @@
...
+
+
+class ScoreNet(ScoreBase):
+
+    def __init__(self, device, cfg, input_dim, output_dim, marginal_prob_std, embed_dim=32):
+        super().__init__(device, cfg, input_dim, output_dim, marginal_prob_std, embed_dim)
+        # The swish activation function
+        self.device = device
+        self.cfg = cfg
+        self.act = lambda x: x * torch.sigmoid(x)
```
Comment (PaParaZz1) [inline ding/model/template/qgpo.py:286]: why not use `SiLU`

File: ding/model/template/qgpo.py [hunk 19]
```diff
@@ -0,0 +1,365 @@
...
+            GaussianFourierProjectionTimeEncoder(embed_dim=embed_dim), nn.Linear(embed_dim, embed_dim)
+        )
+        self.device = device
+        self.noise_schedule = dpm_solver_pytorch.NoiseScheduleVP(schedule='linear')
+        self.dpm_solver = dpm_solver_pytorch.DPM_Solver(
+            self.forward_dmp_wrapper_fn, self.noise_schedule, predict_x0=True
+        )
+        # self.dpm_solver = dpm_solver_pytorch.DPM_Solver(self.forward_dmp_wrapper_fn, self.noise_schedule)
+        self.marginal_prob_std = marginal_prob_std
+        self.q = []
```
Comment (PaParaZz1) [inline ding/model/template/qgpo.py:216]: why do we need list here

File: ding/model/template/qgpo.py [hunk 20]
```diff
@@ -0,0 +1,365 @@
...
+        self.q.append(QGPO_Critic(device, cfg.qgpo_critic, adim=output_dim, sdim=input_dim - output_dim))
+
+    def forward_dmp_wrapper_fn(self, x, t):
```
Comment (PaParaZz1) [inline ding/model/template/qgpo.py:219]: dmp?

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: ding/example/qgpo.py (hunk 8 above)
Comment (zjowowen) [reply ding/example/qgpo.py]: fixed

File: ding/policy/qgpo.py (hunk 9 above)
Comment (zjowowen) [reply ding/policy/qgpo.py]: fixed

File: dizoo/d4rl/config/halfcheetah_medium_expert_qgpo_config.py (hunk 10 above)
Comment (zjowowen) [reply dizoo/d4rl/config/halfcheetah_medium_expert_qgpo_config.py]: fixed

File: ding/torch_utils/network/res_block.py (hunk 11 above)
Comment (zjowowen) [reply ding/torch_utils/network/res_block.py:155]: fixed

File: ding/torch_utils/network/res_block.py (hunk 12 above)
Comment (zjowowen) [reply ding/torch_utils/network/res_block.py]: fixed

File: ding/torch_utils/network/res_block.py (hunk 13 above)
Comment (zjowowen) [reply ding/torch_utils/network/res_block.py]: fixed

File: ding/policy/qgpo.py (hunk 14 above)
Comment (zjowowen) [reply ding/policy/qgpo.py]: fixed

File: ding/policy/qgpo.py (hunk 15 above)
Comment (zjowowen) [reply ding/policy/qgpo.py]: fixed

File: ding/policy/qgpo.py (hunk 16 above)
Comment (zjowowen) [reply ding/policy/qgpo.py]: fixed

File: ding/model/common/encoder.py (hunk 17 above)
Comment (zjowowen) [reply ding/model/common/encoder.py]: fixed

File: ding/model/template/qgpo.py (hunk 18 above)
Comment (zjowowen) [reply ding/model/template/qgpo.py]: fixed

File: ding/model/template/qgpo.py (hunk 19 above)
Comment (zjowowen) [reply ding/model/template/qgpo.py]: fixed

File: ding/model/template/qgpo.py (hunk 20 above)
Comment (zjowowen) [reply ding/model/template/qgpo.py]: fixed

File: ding/example/qgpo.py [hunk 21]
```diff
@@ -0,0 +1,141 @@
...
+        elif reward_tune == 'iql_locomotion':
+            min_ret, max_ret = QGPOD4RLDataset.return_range(data, 1000)
+            reward /= (max_ret - min_ret)
+            reward *= 1000
+        elif reward_tune == 'cql_antmaze':
+            reward = (reward - 0.5) * 4.0
+        elif reward_tune == 'antmaze':
+            reward = (reward - 0.25) * 2.0
+        self.rewards = reward
+        print("dql dataloard loaded")
```
Comment (zjowowen) [inline ding/example/qgpo.py:52]: Change to logging

File: ding/example/qgpo.py [hunk 22]
```diff
@@ -0,0 +1,141 @@
...
+        print(self.len, "data loaded")
+
+    def __getitem__(self, index):
+        data = {
+            's': self.states[index % self.len],
+            'a': self.actions[index % self.len],
+            'r': self.rewards[index % self.len],
+            's_': self.next_states[index % self.len],
+            'd': self.is_finished[index % self.len],
+            'fake_a': self.fake_actions[index % self.len]
```
Comment (zjowowen) [inline ding/example/qgpo.py:64]: add notation for fake data

File: ding/example/qgpo.py [hunk 23]
```diff
@@ -0,0 +1,141 @@
...
+
+        evaluator_env = BaseEnvManagerV2(
+            env_fn=[
+                lambda: DingEnvWrapper(env=gym.make(cfg.env.env_id), cfg=cfg.env, caller="evaluator")
+                for _ in range(cfg.env.evaluator_env_num)
+            ],
+            cfg=cfg.env.manager
+        )
+
+        task.use(qgpo_support_data_generator(cfg, dataset, policy))
```
Comment (zjowowen) [inline ding/example/qgpo.py:113]: try another way.

File: ding/policy/qgpo.py [hunk 24]
```diff
@@ -0,0 +1,261 @@
...
+        s = data['s']
+        a = data['a']
+        r = data['r']
+        s_ = data['s_']
+        d = data['d']
+        fake_a = data['fake_a']
+        fake_a_ = data['fake_a_']
+
+        if self.cuda:
+            s = to_device(s, self._device)
```
Comment (zjowowen) [inline ding/policy/qgpo.py:119]: to device for data

File: ding/policy/qgpo.py [hunk 25]
```diff
@@ -0,0 +1,261 @@
...
+                q0_loss.backward()
+                self.q_optimizer.step()
+
+                # Update target
+                for param, target_param in zip(self._model.q.q0.parameters(), self._model.q.q0_target.parameters()):
+                    target_param.data.copy_(
+                        self.qt_update_momentum * param.data + (1 - self.qt_update_momentum) * target_param.data
+                    )
+
+                q0_loss = q0_loss.detach().cpu().numpy()
```
Comment (zjowowen) [inline ding/policy/qgpo.py:158]: replace detach() by item()

File: ding/policy/qgpo.py [hunk 26]
```diff
@@ -0,0 +1,261 @@
...
+            behavior_model_training_loss = 0
+
+        # training Q function
+        self.energy_guided_policy_begin_training_iter -= 1
+        self.q_value_stop_training_iter -= 1
+        if self.energy_guided_policy_begin_training_iter < 0:
+            if self.q_value_stop_training_iter > 0:
+                q0_loss = self._model.q_loss_fn(a, s, r, s_, d, fake_a_, discount=self.discount)
+
+                self.q_optimizer.zero_grad(set_to_none=True)
```
Comment (zjowowen) [inline ding/policy/qgpo.py:148]: find doc for set_to_none

File: ding/policy/qgpo.py [hunk 27]
```diff
@@ -0,0 +1,261 @@
...
+        Overview:
+            Forward function for eval mode.
+        Arguments:
+            - data (:obj:`dict`): Dict type data.
+        Returns:
+            - output (:obj:`dict`): Dict type data of algorithm output.
+        """
+        data_id = list(data.keys())
+        data = default_collate(list(data.values()))
+        states = data
```
Comment (zjowowen) [inline ding/policy/qgpo.py:216]: delete states

File: ding/policy/qgpo.py [hunk 28]
```diff
@@ -0,0 +1,261 @@
...
+        actions = self._model.select_actions(states, diffusion_steps=self.diffusion_steps)
+        output = actions
+
+        return {i: {"action": d} for i, d in zip(data_id, output)}
+
+    def _get_train_sample(self) -> None:
+        """
```
Comment (zjowowen) [inline ding/policy/qgpo.py:223]: doc for CQL

File: ding/example/qgpo.py (hunk 21 above)
Comment (zjowowen) [reply ding/example/qgpo.py:52]: fixed.

File: ding/example/qgpo.py (hunk 22 above)
Comment (zjowowen) [reply ding/example/qgpo.py:64]: fixed.

File: ding/example/qgpo.py (hunk 23 above)
Comment (zjowowen) [reply ding/example/qgpo.py:113]: fixed.

File: ding/policy/qgpo.py (hunk 24 above)
Comment (zjowowen) [reply ding/policy/qgpo.py:119]: fixed.

File: ding/policy/qgpo.py (hunk 25 above)
Comment (zjowowen) [reply ding/policy/qgpo.py:158]: fixed.

File: ding/policy/qgpo.py (hunk 26 above)
Comment (zjowowen) [reply ding/policy/qgpo.py:148]: fixed.

File: ding/policy/qgpo.py (hunk 27 above)
Comment (zjowowen) [reply ding/policy/qgpo.py:216]: fixed.

File: ding/policy/qgpo.py (hunk 28 above)
Comment (zjowowen) [reply ding/policy/qgpo.py:223]: fixed.

File: ding/framework/middleware/functional/logger.py [hunk 29]
```diff
@@ -611,6 +611,9 @@ def _plot(ctx: "OfflineRLContext"):
             )
 
         if ctx.eval_value != -np.inf:
+            if hasattr(ctx, "info_for_logging"):
+                info_for_logging.update(ctx.info_for_logging)
```
Comment (zjowowen) [inline ding/framework/middleware/functional/logger.py:615]: add comment for it

Comment (zjowowen) [reply ding/framework/middleware/functional/logger.py:615]: fixed.

**Review by PaParaZz1 — APPROVED** [review]

**Pull request merged**